File: leetcode/437.路径总和 III/437.路径总和 III.py
from util.leetcode_type import TreeNode


class Solution:
    def __init__(self):
        self.res = 0

    def pathSum(self, root: TreeNode, targetSum: int) -> int:
        """
        前缀和，回溯算法
        两个节点之差等于 targetSum 即可
        """
        # 记录路径中前缀和出现次数
        # base case 开始是0出现1
        mapping = {0: 1}

        def dfs(root, cur):
            # 结束条件
            if not root:
                return
            # 前缀和+当前节点值
            cur += root.val
            # 判断是否存在符合条件的前缀和
            pre = cur - targetSum
            self.res += mapping.get(pre, 0)
            # 记录当前前缀和
            mapping[cur] = mapping.get(cur, 0) + 1
            dfs(root.left, cur)
            dfs(root.right, cur)
            # 回溯，恢复状态
            mapping[cur] -= 1

        dfs(root, 0)
        return self.res

File: leetcode/437.路径总和 III/437.路径总和 III.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) int {
	/*
		前缀和+dfs回溯
		计算路径上的前缀和，如果有路径和 - 目标值的差值 已经存在，说明有路径和，结果值加上差值的次数
	*/
	// 存储路径和次数， base case：0出现1次
	mapping := map[int]int{0: 1}
	res := 0
	var dfs func(root *TreeNode, cur int)
	dfs = func(root *TreeNode, cur int) {
		if root == nil {
			return
		}
		// 路径和，并记录到哈希表
		cur += root.Val
		mapping[cur] = +1
		// 计算路径和差值，并添加差值出现次数到结果中
		need := cur - targetSum
		res += mapping[need]
		// 递归左右子树
		dfs(root.Left, cur)
		dfs(root.Right, cur)
		// 删除当前路径出现次数
		mapping[cur] -= 1
	}
	dfs(root, 0)
	return res
}
